tests/unit/state/source/test_source_state.cpp

```diff
@@ -1056,3 +1093,764 @@ TEST_F(TestSourceState, SourcesTerminateCorrectly)
+                            kTestBuffer2Length, kDisableForceSync);
+
+  for (size_t i = 0u; i < kTestBufferLength; ++i)
+  {
+    if (i % 2)
+      EXPECT_GT(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+    else
+      EXPECT_EQ(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+  }
+
+  update_levels_and_or_paps(source_state, universe_state, kTestBuffer, kTestBufferLength, nullptr, 0u,
+                            kDisableForceSync);
+
+  for (size_t i = 0u; i < kTestBufferLength; ++i)
+  {
+    if (i % 2)
+      EXPECT_GT(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+    else
+      EXPECT_EQ(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+  }
+
+  disable_pap_data(universe_state);
+  update_levels_and_or_paps(source_state, universe_state, kTestBuffer, kTestBufferLength, nullptr, 0u,
+                            kDisableForceSync);
+
+  for (size_t i = 0u; i < kTestBufferLength; ++i)
+    EXPECT_GT(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+
+  update_levels_and_or_paps(source_state, universe_state, kTestBuffer, kTestBufferLength, &kTestPriority, 1u,
+                            kDisableForceSync);
+
+  for (size_t i = 0u; i < kTestBufferLength; ++i)
+  {
+    if (i == 0u)
+      EXPECT_GT(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+    else
+      EXPECT_EQ(universe_state->level_send_buf[SACN_DATA_HEADER_SIZE + i], 0u);
+  }
+}
+
+TEST_F(TestSourceState, UpdateLevelsIncrementsActiveUniversesCorrectly)
+{
+  sacn_source_t source = AddSource(kTestSourceConfig);
+  uint16_t universe = AddUniverse(source, kTestUniverseConfig);
+
+  SacnSource* source_state = nullptr;
+  SacnSourceUniverse* universe_state = nullptr;
+  lookup_source_and_universe(source, universe, &source_state, &universe_state);
+
+  EXPECT_EQ(source_state->num_active_universes, 0u);
+  update_levels_and_or_paps(source_state, universe_state, kTestBuffer, kTestBufferLength, nullptr, 0u,
+                            kDisableForceSync);
+  EXPECT_EQ(source_state->num_active_universes, 1u);
+  update_levels_and_or_paps(source_state, universe_state, kTestBuffer2, kTestBuffer2Length, nullptr, 0u,
+                            kDisableForceSync);
+  EXPECT_EQ(source_state->num_active_universes, 1u);
+
+  SacnSourceUniverseConfig unicast_only_config = kTestUniverseConfig;
+  ++unicast_only_config.universe;
+  unicast_only_config.send_unicast_only = true;
+  uint16_t unicast_only_universe = AddUniverse(source, unicast_only_config);
+  SacnSourceUniverse* unicast_only_universe_state = nullptr;
+  lookup_source_and_universe(source, unicast_only_universe, &source_state, &unicast_only_universe_state);
+
+  update_levels_and_or_paps(source_state, unicast_only_universe_state, kTestBuffer, kTestBufferLength, nullptr, 0u,
+                            kDisableForceSync);
+  EXPECT_EQ(source_state->num_active_universes, 1u);
+}
+
+TEST_F(TestSourceState, IncrementSequenceNumberWorks)
+{
```

Comment:
We need a version of this test that tests for proper rollover.

Reply:
Great idea! I'll include that in the next PR.